pywavefront/parser.py

```diff
@@ -31,44 +31,136 @@
 # ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 # ----------------------------------------------------------------------------
-import os
+import codecs
+import gzip
+import io
 import logging
+import os
+import sys
+
+from pywavefront.exceptions import PywavefrontException
+
+
+def auto_consume(func):
+    """Decorator for auto consuming lines when leaving the function"""
+    def inner(*args, **kwargs):
+        func(*args, **kwargs)
+        args[0].consume_line()
+    return inner
 
 
 class Parser(object):
     """This defines a generalized parse dispatcher; all parse functions
     reside in subclasses."""
-    strict = False
 
-    def __init__(self, file_name):
+    def __init__(self, file_name, strict=False, encoding="utf-8"):
+        """
+        Initializer for the base parser
+        :param file_name: Name and path of the file to read
+        :param strict: Enable or disable strict mode
+        """
         self.file_name = file_name
+        self.strict = strict
+        self.encoding = encoding
         self.dir = os.path.dirname(file_name)
 
-    def read_file(self):
-        with open(self.file_name, 'r') as file:
-            for line in file:
-                self.parse(line)
-
-    def parse(self, line):
-        """Determine what type of line we are and dispatch
-        appropriately."""
-        if line.startswith('#'):
-            return
-
-        values = line.split()
-        if len(values) < 2:
-            return
-
-        line_type = values[0]
-        args = values[1:]
-        attrib = 'parse_%s' % line_type
-
-        if Parser.strict:
-            parse_function = getattr(self, attrib)
-            parse_function(args)
-        elif hasattr(self, attrib):
-            parse_function = getattr(self, attrib)
-            parse_function(args)
+        self.dispatcher = self._build_dispatch_map()
+        self.lines = self.create_line_generator()
+
+        self.line = None
+        self.values = None
+
+    def create_line_generator(self):
+        """
+        Creates a generator function yielding lines in the file
+        Should only yield non-empty lines
+        """
+        if self.file_name.endswith(".gz"):
+            if sys.version_info.major == 3:
+                gz = gzip.open(self.file_name, mode='rt', encoding=self.encoding)
+                f = io.BufferedReader(gz)
```

Comment:
It looks like `f` is not used in this block.

Reply:
Good catch.